parla/task_runtime.py

```diff
@@ -525,10 +540,20 @@ def run(self):
                 logger.exception("Exception in task")
             finally:
                 logger.info("Finally for task %r", self)
-                # Deallocate all the resources, both from the allocation above and from the "assignment" done by
-                # the scheduler.
+
+                # Deallocate resources
                 for d in self.req.devices:
+                    for resource, amount in self.req.resources.items():
+                        logger.debug("Task %r deallocating %d %s from device %r", self, amount, resource, d)
                     ctx.scheduler._available_resources.deallocate_resources(d, self.req.resources)
+                    ctx.scheduler._device_compute_task_counts[d] -= 1
+
+                # Update parray tracking information
+                # The easiest way to do this is just untrack and re-track the array for anything that could've changed
+                for parray in (self.dataflow.inout + self.dataflow.output):
+                    ctx.scheduler._available_resources.untrack_parray(parray)
+                    ctx.scheduler._available_resources.track_parray(parray)
```

Comment:
Would there be any problems if the GIL switches to another thread/task between 554 and 555?

Reply:
aaaah this is making me realize i have several mutexes where GIL switching could mess things up

Reply:
I'm adding it as a TODO. I think it's unlikely to bite us rn (knock on wood) and it's something I can look at fixing later when I'm also fixing evictions to free memory.